src/lib/operators/join_hash.cpp

```diff
@@ -74,41 +74,41 @@ void JoinHash::_on_set_parameters(const std::unordered_map<ParameterID, AllTypeV
 template <typename T>
 size_t JoinHash::calculate_radix_bits(const size_t build_relation_size, const size_t probe_relation_size) {
   /*
-    Setting number of bits for radix clustering:
-    The number of bits is used to create probe partitions with a size that can
-    be expected to fit into the L2 cache.
-    This should incorporate hardware knowledge, once available in Hyrise.
-    As of now, we assume a L2 cache size of 1024 KB (L2 cache size of recent
-    Intel Xeon CPUs), of which we use 50%.
+    The number of radix bits is used to determine the number of build partitions whose hash maps have a size that can
+    be expected to fit into the L2 cache. This should incorporate hardware knowledge, once available in Hyrise. As of
+    now, we assume a L2 cache size of 1024 KB, of which we use 75 %.
+
     We estimate the size the following way:
       - we assume each key appears once (that is an overestimation space-wise, but we
       aim rather for a hash map that is slightly smaller than L2 than slightly larger)
-      - each entry in the hash map is a uint32_t offset (see hash_join_steps.hpp)
+      - each entry in the hash map is a pair of the actual hash key and the SmallPosList storing uint32_t offsets (see
+        hash_join_steps.hpp)
   */
   if (build_relation_size > probe_relation_size) {
     /*
-      Hash joins perform best when the build relation is small. In case the
-      optimizer selects the hash join due to such a situation, but neglects that the
-      input will be switched (e.g., due to the join mode), the user will be warned.
+      Hash joins perform best when the build relation is small. In case the inputs are switched (e.g., due to the join
+      mode) making the build partition larger than the probe partition, the user will be warned.
```

Comment:
I read that this is only for CI, but just thought of this:

This sounds as if there was a right and a wrong way for the optimizer to specify the join. Someone new might try to find the code that is responsible for avoiding the flipping.

How do you feel about getting rid of the "flip" terminology altogether and simply reporting whether the left or the right side is the build side?

Reply:
Maybe something we should discuss on Wednesday.

Reply:
Just notes for Friday. Handling this in the optimizer is something I would like to work on, but I don't have the time right now. It's a simple heuristic. Nothing fancy.

The problem is that the only alternative is the sort-merge join and we have broke that join's performance with the introduction of the secondary attributes.

Reply:
Could you remind me why we even care about this? I'm having a hard time remembering in which cases this makes a difference.

Reply:
Mainly because hash joins suck when the build side is much larger. Even right now with the sub-optimal sort-merge join we could easily speed up several of these joins.

Reply:
My point was that I wonder why the optimizer has to deal with it.

Reply:
I don't get where this question comes from. There are no changes to the optimizer.

The place that I would currently change is rather the LQP translator.

Reply:
The LQPTranslator should not touch any statistics. Even if - what advantage do we have if this is done in the LQP rather than in the operator?

Reply:
I don't plan to touch neither the LQP translator, nor optimizer, nor the internals of the join hash. I don't see where this discussion is going.

If statistics are needed, then yes, I should not end up in the LQP translator. In this case, I'd vote for a LQPNode marker similar to the PredicateNode.
My personal favorite would be a proxy join operator that decides during runtime, but it wasn't received very well back in the days when we first built that with @SvenLehmann.

Does that answer your question? Maybe discuss via phone?

Reply:
Let's do it just before you pick this PR up again.

Reply:
I misunderstood your first comment and totally lost track of what we discussed afterwards.

Changed the output for the visualization to` "Build side is " << (left_input_is_build_side ? "left." : "right.");"`. So no mention of "flip".

Do you know what we have discussed afterwards?

Reply:
Hmm, I can't recall all of it. From looking at the previous discussion and the code, I think I was missing a distinction between two cases:

1. The optimizer passes the tables so that the build table is on the left. I still don't understand how that is any issue at all. Previously, we said that we "flip" the table. Instead, we should go for "build"/"probe" side as early as possible. That might already be the case. 
2. In the second case, we had to pick the larger table as the probe table. This is the case where we should show a warning, saying that the hash join might not be the most efficient join operator for this. The text and the comment about it should be clearer. Not just "e.g., due to the join mode". As far as I understand it, this is not an example but the only case where this warning should be printed. We should explain that certain join modes require the a certain table to be the build table - even if it is bigger.

Also, is "build partition" really the correct term here? Have we already partitioned the input? Also, do we need to talk about relations? I always find that term unnecessarily academic if everyone knows that we are talking about tables.